src/main/java/sparksoniq/spark/udf/GroupClauseCreateColumnsUDF.java

```diff
@@ -84,37 +88,52 @@ public Row call(WrappedArray wrappedParameters) {
             expression.open(_context);
             boolean isEmptySequence = true;
-            while (expression.hasNext()) {
+            if (expression.hasNext()) {
                 isEmptySequence = false;
                 Item nextItem = expression.next();
-                if (nextItem instanceof NullItem) {
+                if (nextItem.isNull()) {
                     _results.add(nullGroupIndex);
-                    _results.add(null);     // placeholder for valueColumn(2nd column)
-                } else {
-                    // any other atomic type
-                    _results.add(valueGroupIndex);
-
-                    // extract type information for the sorting column
-                    String typeName = (String) _allColumnTypes.get(expressionIndex);
-
-                    if (typeName.equals("bool")) {
-                        _results.add(nextItem.getBooleanValue());
-                    } else if (typeName.equals("string")) {
-                        _results.add(nextItem.getStringValue());
-                    } else if (typeName.equals("integer")) {
-                        _results.add(Item.getNumericValue(nextItem, Integer.class));
-                    } else if (typeName.equals("double")) {
-                        _results.add(Item.getNumericValue(nextItem, Double.class));
-                    } else if (typeName.equals("decimal")) {
-                        _results.add(Item.getNumericValue(nextItem, BigDecimal.class));
+                    _results.add(null);
+                    _results.add(null);
+                } else if (nextItem.isBoolean() ){
+                    if(nextItem.getBooleanValue())
+                    {
+                        _results.add(trueGroupIndex);
                     } else {
-                        throw new SparksoniqRuntimeException("Unexpected grouping type found while creating columns.");
-                    }
+                        _results.add(falseGroupIndex);
+                    }                        
+                    _results.add(null);
+                    _results.add(null);
+                } else if (nextItem.isString()) {
+                    _results.add(stringGroupIndex);
+                    _results.add(nextItem.getStringValue());
+                    _results.add(null);
+                } else if (nextItem.isInteger()) {
+                    // any other atomic type
+                    _results.add(doubleGroupIndex);
+                    _results.add(null);
+                    _results.add(new Double(nextItem.getIntegerValue()));
+                } else if (nextItem.isDecimal()) {
+                    // any other atomic type
+                    _results.add(doubleGroupIndex);
+                    _results.add(null);
+                    _results.add(new Double(nextItem.getDecimalValue().doubleValue()));
+                } else if (nextItem.isDouble()) {
+                    // any other atomic type
```

Comment:
The any other atomic type comment is repeated here and seems redundant. Exception is thrown for the non-atomic case below.